torch/lib/TH/generic/THTensorMath.c

```diff
@@ -349,6 +349,81 @@ void THTensor_(indexCopy)(THTensor *tensor, int dim, THLongTensor *index, THTens
   THLongTensor_free(index);
 }
 
+static ptrdiff_t THTensor_(dataOffset)(THTensor* tensor, ptrdiff_t linearIndex) {
+  int64_t *size = tensor->size;
+  int64_t *stride = tensor->stride;
+  int nDim = tensor->nDimension;
+  ptrdiff_t dataOffset = 0;
+  for (int i = nDim - 1; i >= 0; i--) {
+    dataOffset += (linearIndex % size[i]) * stride[i];
+    linearIndex /= size[i];
+  }
+  return dataOffset;
+}
+
+static int64_t THTensor_(wrapLinearIndex)(int64_t linearIndex, int64_t numel) {
+  THArgCheck(linearIndex < numel && linearIndex >= -numel, 2, "out of range");
+  return linearIndex < 0 ? linearIndex + numel : linearIndex;
+}
+
+void THTensor_(take)(THTensor *r_, THTensor *src, THLongTensor *index)
+{
+  THTensor_(resizeNd)(r_, index->nDimension, index->size, NULL);
+  THTensor* dst = THTensor_(newContiguous)(r_);
```

Comment:
Doesn't `resizeNd` ensure that `r_` is contiguous?

Reply:
I thought a bit about this and I'm unsure whether we should change it or not. Consider a call like `torch.add(x, 1, out=y[0])`. Right now, if `y[0]` has a correct size, it will be filled with `x + 1`, otherwise it will be a temporary tensor that will get resized and will no longer share storage with `y`. We should do sth about it...

Reply:
I agree it would be good to handle this. I found an example with zero-strided output tensors that is another example of something that has bugged me in the past https://github.com/torch/torch7/issues/289

Reply:
@apaszke This feels related to #3313. Basically, if we write data into a tensor as an out, this should count as an "inplace" operation (even though its not actually inplace).